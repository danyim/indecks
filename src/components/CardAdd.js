import React from 'react';
import { Link, browserHistory } from 'react-router';
import styles from '../styles/components/CardAdd';

class CardAdd extends React.Component {
  constructor(props) {
    super(props);

    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleSubmit(e, deckId) {
    e.preventDefault();
    const card = {
      title: this.refs.title.value,
      answer: this.refs.answer.value
    };
    this.props.handleSubmit(deckId, card);
  }

  render() {
    const { deckId } = this.props;

    return (
      <figure className={`grid-figure ${styles['grid-figure']}`}>
        <form ref="commentForm" className="edit-form" onSubmit={e => this.handleSubmit(e, deckId)}>
          <input type="text" className="large-input" name="title" ref="title" placeholder="Title" />
          <textarea type="text" className="mono" name="answer" ref="answer" placeholder="Answer (Markdown)" rows="4" />
          <button type="submit" className="button">Save</button>
        </form>
      </figure>
    );
  }
}

CardAdd.defaultProps = {};
CardAdd.propTypes = {
  deckId: React.PropTypes.string.isRequired,
  handleSubmit: React.PropTypes.func.isRequired
};

export default CardAdd;
