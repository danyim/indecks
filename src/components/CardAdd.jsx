import React from 'react';
import { findDOMNode } from 'react-dom';
import { Link, browserHistory } from 'react-router';
import styles from '../styles/components/CardAdd';

const propTypes = {
  deckId: React.PropTypes.string.isRequired,
  handleSubmit: React.PropTypes.func.isRequired
};

const defaultProps = {};

class CardAdd extends React.Component {
  constructor(props) {
    super(props);

    this.handleCancel = this.handleCancel.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  componentDidMount(){
    // This happens a little too fast. If you get to this view via keyboard
    // shortcut, the focus event fires before the keyup and writes the value
    // into the field. Adding a 100ms delay prevents this from happening.
    setTimeout(() => {
      findDOMNode(this.refs.title).focus()
    }, 100);
  }

  handleSubmit(e, deckId) {
    e.preventDefault();
    const card = {
      title: this.refs.title.value,
      answer: this.refs.answer.value
    };
    this.props.handleSubmit(deckId, card);
  }

  handleCancel() {
    this.props.history.goBack();
  }

  render() {
    const { deckId } = this.props;

    return (
      <figure className={`grid-figure ${styles['grid-figure']}`}>
        <form ref="commentForm" className="edit-form" onSubmit={e => this.handleSubmit(e, deckId)}>
          <input type="text" className="large-input" name="title" ref="title" placeholder="Title" />
          <textarea type="text" className="mono" name="answer" ref="answer" placeholder="Answer (Markdown)" rows="4" />

          <div className={`${styles['control-buttons']}`}>
            <button type="submit" className="button">Save</button>
            <button type="button" className="button" onClick={this.handleCancel}>Cancel</button>
          </div>
        </form>
      </figure>
    );
  }
}

CardAdd.propTypes = propTypes;
CardAdd.defaultProps = defaultProps;

export default CardAdd;
