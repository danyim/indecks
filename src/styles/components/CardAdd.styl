@import '../_variables'

figure.grid-figure
  display flex
  min-height 275px
  flex-flow column nowrap
  justify-content: center

.control-buttons
  flex: 0 0 auto
  flex-flow: row nowrap
  button
    margin-left 1rem
  button:first-child
    margin-left 0rem
