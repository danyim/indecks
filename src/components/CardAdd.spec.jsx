import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import CardAdd from './CardAdd';

const defaultProps = {
  deckId: 'ABCDEFG',
  handleSubmit: () => {},
  history: {
    goBack: () => {}
  }
};

function setup(props = defaultProps) {
  const wrapper = shallow(<CardAdd {...props} />)

  return {
    props,
    wrapper
  }
}

describe('CardAdd', () => {
  it('should render self and subcomponents', () => {
    const { wrapper } = setup();
    expect(wrapper.find('figure.grid-figure').exists()).toBe(true);

    const tree = renderer.create(
      <CardAdd {...defaultProps} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should call the handleSubmit prop when submitting the form', () => {
    const handler = jest.fn();
    const { wrapper } = setup({
      ...defaultProps,
      handleSubmit: handler
    });

    const title = wrapper.find('input[name="title"]');
    const answer = wrapper.find('textarea[name="answer"]');
    title.simulate('change', { target: { value: 'Test title' } });
    answer.simulate('change', { target: { value: 'Test description' } });

    wrapper.find('form').simulate('submit', {
      preventDefault: () => {}
    });

    expect(handler.mock.calls.length).toBe(1);
  });

  it('should not allow a blank title', () => {
    const handler = jest.fn();
    const { wrapper } = setup({
      ...defaultProps,
      handleSubmit: handler
    });

    const title = wrapper.find('input[name="title"]');
    const answer = wrapper.find('textarea[name="answer"]');
    title.simulate('change', { target: { value: '' } });
    answer.simulate('change', { target: { value: 'Test' } });

    wrapper.find('form').simulate('submit', {
      preventDefault: () => {}
    });

    expect(handler.mock.calls.length).toBe(0);
  });

  it('should not allow a blank answer', () => {
    const handler = jest.fn();
    const { wrapper } = setup({
      ...defaultProps,
      handleSubmit: handler
    });

    const title = wrapper.find('input[name="title"]');
    const answer = wrapper.find('textarea[name="answer"]');
    title.simulate('change', { target: { value: 'Test' } });
    answer.simulate('change', { target: { value: '' } });

    wrapper.find('form').simulate('submit', {
      preventDefault: () => {}
    });

    expect(handler.mock.calls.length).toBe(0);
  });

  it('should navigate backwards in the browser history when cancel is clicked', () => {
    const handler = jest.fn();
    const { wrapper } = setup({
      ...defaultProps,
      history: {
        goBack: handler
      }
    });

    const cancel = wrapper.find('button[children="Cancel"]');
    cancel.simulate('click');

    expect(handler.mock.calls.length).toBe(1);
  });
});
