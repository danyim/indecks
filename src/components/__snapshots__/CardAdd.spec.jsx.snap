// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardAdd should render self and subcomponents 1`] = `
<CardAdd
  deckId="ABCDEFG"
  handleSubmit={[Function]}
  placeholderGenerator={[Function]}
  push={[Function]}
>
  <figure
    className="grid-figure grid-figure"
  >
    <form
      className="edit-form"
      onSubmit={[Function]}
    >
      <label
        htmlFor="title"
      >
        <span>
          Card Title
        </span>
        <input
          className="large-input"
          name="title"
          onChange={[Function]}
          placeholder="Card title here..."
          tabIndex={0}
          type="text"
        />
      </label>
      <label
        htmlFor="answer"
      >
        <span>
          Card Answer  (
          <a
            href="//guides.github.com/features/mastering-markdown/"
            rel="noreferrer noopener"
            target="_blank"
          >
            Markdown
          </a>
          )
        </span>
        <textarea
          className="mono"
          name="answer"
          onChange={[Function]}
          placeholder="Card answer here (supports Markdown)..."
          rows="4"
          tabIndex={-1}
          type="text"
        />
      </label>
      <CardPreview
        answer=""
        title=""
      >
        <div
          className="preview"
        >
          <button
            className="btn pointer m-t"
            name="preview-toggle"
            onClick={[Function]}
            type="button"
          >
            Preview [
            +
            ]
          </button>
          <Motion
            style={
              Object {
                "y": Object {
                  "damping": 26,
                  "precision": 0.01,
                  "stiffness": 170,
                  "val": 0,
                },
              }
            }
          >
            <div
              className="preview-pane"
              style={
                Object {
                  "display": "none",
                  "height": 0,
                }
              }
            >
              <label
                className="no-pad"
                htmlFor="title"
              >
                <span>
                  Title
                </span>
                <Markdown
                  className="card-title"
                  handleOnClick={null}
                  name="title"
                  text=""
                >
                  <div
                    className="card-title"
                    dangerouslySetInnerHTML={
                      Object {
                        "__html": "",
                      }
                    }
                    onClick={null}
                  />
                </Markdown>
              </label>
              <hr />
              <label
                className="no-pad"
                htmlFor="answer"
              >
                <span>
                  Answer
                </span>
                <Markdown
                  className="card-answer"
                  handleOnClick={null}
                  name="answer"
                  text=""
                >
                  <div
                    className="card-answer"
                    dangerouslySetInnerHTML={
                      Object {
                        "__html": "",
                      }
                    }
                    onClick={null}
                  />
                </Markdown>
              </label>
            </div>
          </Motion>
        </div>
      </CardPreview>
      <div
        className="control-buttons"
      >
        <button
          className="button outline"
          type="submit"
        >
          Save
        </button>
        <button
          className="button"
          onClick={[Function]}
          type="button"
        >
          Cancel
        </button>
      </div>
    </form>
  </figure>
</CardAdd>
`;
